### ISSUE TITLE
Develop a Roadmap

### ISSUE THREAD

**Issue by scopatz** [dask/dask#3589]
Hello All, 

I'd like to help dask develop a roadmap, and @mrocklin requested that I open an issue here to kickstart the discussion. 

For those of you unfamiliar, a roadmap is a listing of near- and long-term goals that the project has, but which have yet to be implemented. These goals are generally larger than single pull request. "Add GPU Support" or "Port to Python 3" are examples of some roadmap items a project might have. However, goals can also include activities that seem small and mundane, but are critical to the project, such as "Improve Documentation" or "Achieve 100% Test Coverage."

**The purpose of this issue is so that roadmap items can be listed and commented on in the discussion.** Once we have a reasonable sense of what folks would like to see on the roadmap, the contents should make its way to the website/docs in some fashion. It is also a good idea if it ends up being listed in priority order, so that higher priority items are closer to the top. 

Some good example of project roadmaps are:

* [SymPy Roadmap](http://sympy.org/en/roadmap.html)
* [Bokeh Roadmap](https://bokehplots.com/pages/roadmap.html)

A decent example of a project roadmap is [Spyder's Roadmap](https://github.com/spyder-ide/spyder/wiki/Roadmap) which looks more like a timeline with milestones.

Roadmaps serve a joint purpose:

* They help inform and attract potentail funders, and some funders require roadmap documents
* They help attract contributors, as developers are more likely to jump in an help an establish project than start a new project to scratch their itch.

**TL;DR:** Roadmaps communicate a project's intentions. It would be a good idea for dask to have one, and I am happy to help facilitate!

So where would you like to see dask go?

**Comment by scopatz** [comment]
@mrocklin - yeah, I hear you about the mixed use here.  Option 1 is clearly a contractual roadmap, "We plan to do X. Time, materials, and hands speed up the process of achieving X."  Option2 is more of a wishlist roadmap, "Given time, materials, and hands, we'd love to do X and are currently accepting contributions."

I think terms like "wishlist," "prospectus", "stretch goals," and "desired milestones" all fit the bill for option 2, if you would like alternative verbiage. 

However, I am not exclusively describing option 2. I believe, that for most projects, it is tempting for the two cases you have described to make their way into the same document. This is because folks approach the roadmap question as, "What would I like the project to look like if I had infinite resources?" So every crazy idea that has ever popped into the developers heads get listed initially. These infitine resource ideas tend to get weeded out (or at least pushed down) during prioritization. 

Rather, I am suggesting whatever the you all think is approriate for a roadmap document. If you believe it should be more contractual, then option 1 alone seems likt the correct path. If you'd like some of the more whishful ideas to make it in, they can certainly be included either:

a. Intermixed with the contractual items, or
b. Clearly deliniated into their own section so as not to be confused.

These roadmaps are reflections of the projects and the authors, and so it is really up to the developers as to how structure such documents.

In the situation where there are both contractual and wishlist elements, I am personally less interested in seeing the wishlist side of things. But that doesn't mean that you should or should not add them. There aren't really wrong answers here!

**Comment by mrocklin** [comment]
To be clear, I don't think that this project has the ability to write a contractual roadmap.  This assumes that the project has dedicated long-term resources and autonomy, which it doesn't.  

Currently I answer "what is your roadmap?" questions by saying what current funding sources and what current volunteer developer communities care about.

>  I am personally less interested in seeing the wishlist side of things

Interesting, I was actually assuming that you were asking for the wishlist side of things.

If I were to write such a document on my own (which I have no particular desire to do) I would probably phrase it as a set of projects that are well-scoped and that people could take on in finite time.  This would likely be somewhat similar to [this document for summer internship projects](http://matthewrocklin.com/blog/work/2018/03/20/summer-projects) though aimed towards larger endeavors.

**Comment by eriknw** [comment]
Yeah, roadmap to me implies a schedule, a sequence of milestones, or list of priorities.  Interestingly, this is not what the SymPy and Bokeh roadmaps that you linked are like.  They are more like strategic goals.  They could be perceived as what the primary developers intend to work on.

For dask, it may be helpful to list or tag outstanding PRs and issues that have the most strategic importance, but are not actively being worked on.  For example multi-index for dask.dataframe.  This has the advantage of using existing procedures (github issues and tags), and may more-or-less satisfy the two purposes @scopatz listed.

Perhaps adding a "Vision" section to the docs would be more useful than adding a "Roadmap" section.  However, I'm not sure how different this would be from the "Why dask?" section.

Pieter Hintjens' take on roadmaps for ZeroMQ is pretty interesting:
http://zguide.zeromq.org/page:all#How-ZeroMQ-Lost-Its-Road-Map

**Comment by scopatz** [comment]
@eriknw Thanks for the link! I think the issue with ZeroMQ is that they attached dates to the roadmap item. I don't think it is a good idea to promise both features (or other deliverables) and dates when they will be delivered, for exactly the reason Pieter Hintjen says. An all volunteer, open source workforce is simply not reliable enough.

A "Vision" section would be great.  I do think this is distinct from the "Why Dask?" section, which tries to answer, "Why dask now?" A vision section would instead answer, "What is the future of dask, and how can I help improve it (via code, funds, etc)?"

**Comment by scopatz** [comment]
Pinging this! Are there still questions or concerns about the format or structure that I can help answer? FWIW, I think that what Dask-ML has done for their roadmap has hit the nail on the head.

**Comment by mrocklin** [comment]
I think that my concerns from https://github.com/dask/dask/issues/3589#issuecomment-396321758 still hold.  From the discussion above I think that there is still some confusion about what you're trying to accomplish.  My original guess was that your direct mandate was the following:

> Find features that some other company or funding agency might see and request to be implemented, so that a company like QuanSight might perform that work.

I'm definitely happy to find things like this.  I'm -1 on calling that wishlist of features a roadmap, which I believe offers a promise that we're actually going towards those things.  Your statement that you don't want a wishlist of features leaves me confused about the objective here.

If you're looking for a list of features that Dask developers plan to implement then that list only has a horizon of a few months, and so probably isn't worth keeping up to date.

**Comment by scopatz** [comment]
So I think there are a couple of points of confusion that hopefully I can clear up. 

> Your statement that you don't want a wishlist of features leaves me confused about the objective here.

First, I would like to see a wish list of features. But in the discussion above we had set up a dichotomy between more readily achivable features (which we'd called contractual) and more blue sky ones (wish list). My claim above was intended to say that I personally am much more interested in more achievable features than less achievable ones. Also, drawing out my desire to see more achievable features as a personal preference was intended to relate that it should have no bearing on what dask chooses to put in a wish list; go full blue sky if you want! This is your wishlist, not mine. 

> If you're looking for a list of features that Dask developers plan to implement then that list only has a horizon of a few months, and so probably isn't worth keeping up to date.

I think planning items such as this are also fair game, and generally of interest to your user community. These kinds of feature listings I usually consider living documents. To that end, part of the longer term plan is to help update this document every 6 - 12 months. However, if that is not how you want to view this exercise, that is fine too! Again, this is Dask's, and it should fit the needs of the Dask community. 

> My original guess was that your direct mandate was the following:

>> Find features that some other company or funding agency might see and request to be implemented, so that a company like QuanSight might perform that work.

So this isn't quite what my mandate is. At this step it really is just the altruistic, "help open source project develop roadmaps / wishlists / prospecti / etc."  This is one of the main reasons I joined Quansight. I believe that open source projects should have roadmaps in the same way that they have licenses, contributor guidlines, governance documents, etc. I think that serve an important and oft overlooked communication function. This should be viewed in the context of open source sustainability. 

I also know that these are notoriously difficult to write unless someone external is proding a project to do so. I am definitely personally guilty of not having roadmaps or wishlists for some of the major projects I am involved with (e.g. xonsh, conda-forge). So one of my hats at Quansight is to increase the number of projects out there with roadmaps. I am happy that I now have the oppurtunity to help work on this issue of roadmaps/wishlists more broadly, which has been bothering me for a few years (since SymPy put out thier roadmap).

That said, there is a separate follow-on marketing effort to produce promotional material (free for anyone to use) based off of the roadmap / wishlist. My role in that effort is only to communicate with the developer community and ensure that they have a strong voice in the promotional material that is created. We wouldn't put out anything against the wishes of the project. 

I hope this helps clarifies some of the blocking issues here. As always, I am happy to continue answering questions or concerns about the process, goals, or just where I am coming from on this topic!

**Comment by mrocklin** [comment]
That does help.  

> To that end, part of the longer term plan is to help update this document every 6 - 12 months.

To be clear, when I say "we have a horizon of a few months" I think that that's really more like 2-3.  I think that the right way to communicate what people are working on at this time frame is through things like blogposts and meeting minutes rather than roadmap documents.  I don't know what people within the Dask organization are likely to be working on in six months (though they themselves might) so I'm not sure how I personally can make progress here.

Again, I'm happy to talk about "things that people might work on that would maybe be beneficial and interesting" but most of these are going to be external anyway, like "use dask to parallelize geopandas" or "consider how (in)effective a dynamic system like Dask is for numerical linear algebra" but those really depend more on the interests of those involved than on a central authority of the project.

By the way, if any other Dask developers want to join in please do.  I'm hesitant here, but others may have more optimistic thoughts.

**Comment by scopatz** [comment]
Hello again, I am hoping we can reframe this conversation into more of a "wishlist."  That is, a list of features people have requested, including you all as developers. 

Even if you could just bullet point the ideas here in this issue, that would be enough for us to move forward.

Thanks! I'd really appreciate it!

**Comment by arm2arm** [comment]
dask distributed data access without client:
1) client doesnot have an access to the filesystem but  scheduler and workers.
this is the case when the data is located in the cluster environment, but client notebook is running on a laptop :)

2) cluster nodes got  some  local fast nvmes but login node and scheduler not 
would be cool to have: worker1-100:file://nvme/data-0-100
I know there are some design problems that client needs to know metadata,number of files etc,but again  this would be a nice way to deal with huge datasets on the remote cluster.

**Comment by martindurant** [comment]
In the context of [fsspec](https://github.com/martindurant/filesystem_spec), which I have been thinking about a lot lately, I think it would be possible to any filesystems seen by workers as yet another filesystem implementation. This would take a little work do make happen, but could be a possible mid-term solution.

**Comment by arm2arm** [comment]
an another idea on shared data between users:
different clients are sharing readonly workers DataFrames. so one can load  the data  into the memory and ask different questions over the different clients. something like distributed database?

**Comment by bryevdv** [comment]
FYI the original Bokeh link above is out of date/irrelevant. We now make this distinction explicit, listing Areas of Work that would be in scope, contingent on resources. https://bokeh.pydata.org/en/latest/docs/vision.html#areas-of-work

**Comment by rgommers** [comment]
This wish list or "areas of work" or whatever you want to call it would indeed be very helpful to have. Most helpful are the things that are _not_ already in the pipeline and are likely to be delivered within 2-3 months. Bokeh's doc is a good example, and so is [SciPy's](http://scipy.github.io/devdocs/roadmap.html). Those are things that many people would really like to have, but are in many cases not moving forward (or not as fast as we'd like), because of lack of developer time.

Try to formulate an answer to this question: if I gave you a few talented engineers with as the only constraint they have to work on something more specific than "general maintenance and code review", what would you let them work on?

This is a useful doc to attempt to (a) get more effort going into Dask, and (b) get that work done on the right things. Concrete example: say that my company employs a Dask maintainer. And that person would like to spend more work time on Dask. That's awesome, but at some point I'll need to find funding for that. Obtaining funding for "general maintenance" or "whatever that maintainer is interested in" is _very_ difficult. Funders want to know what they're funding. It's equally useful for things like building a shared understanding of project direction, or for attracting new contributors and giving them a more complete picture of the project.

> Again, I'm happy to talk about "things that people might work on that would maybe be beneficial and interesting" but most of these are going to be external anyway, like "use dask to parallelize geopandas" 

I don't understand this statement. Are you saying Dask itself is basically done?

**Comment by mrocklin** [comment]
> I don't understand this statement. Are you saying Dask itself is basically done?

Relative to its use, yes.  I think that outreach or applying Dask to other domains is likely much more important than improving Dask itself.  If I had a free engineer, I would tell them to go explore a particular domain (indeed, this is my current situation).  So today my list would be something like:

1.  Use Dask in imaging
2.  Use Dask is radio astronomy
3.  Use Dask to build parameter servers
4.  Use Dask on GPUs
5.  Use Dask in GIS
6.  Use Dask in genomics
7.  See if Numba and Dask together can solve PDEs, where does this break down?

We did this with the Pangeo group, and it uncovered a ton of work to do that was highly pragmatic both for the project, and for the scientific discipline.  We also found that it was very easy to fund this work from government grants.  I would strongly encourage this model into the future. 

If you want a list of improvements in Dask, I recommend talking to user groups like Pangeo.  They'll be able to provide a wishlist for you that is specific to their domain, and likely relevant for others (cc @rabernat).

**Comment by rgommers** [comment]
> Use Dask to build parameter servers
> Use Dask on GPUs

Both of those are not other domains, they're major items for your wish list.

> uncovered a ton of work to do that was highly pragmatic both for the project

maybe you did it all, but uncovering a ton of work is exactly the point where you put some work items on your wish list.

> See if Numba and Dask together can solve PDEs

What then goes on your wish list is "improve interaction between Dask and Numba". 

> If you want a list of improvements in Dask, I recommend talking to user groups like Pangeo.

I'm suggesting _the Dask team_ does this, and writes down the outcomes. It's not hard, you already have talked to them a lot, and probably are aware of some things they want. Why let me reinvent the wheel?

Anyway, I already have a partial answer here. Start of a wish list:

- Implement a parameter server for distributed training in dask-ml (see https://github.com/dask/dask-ml/issues/171)
- Improve Dask usability and performance on GPUs
- Improve interaction between Dask and Numba

All of those things are perfectly understandable and great possible conversation starters with funders. If you just write down a few more items like this and put them on wiki or in the docs, this issue can be closed:)

**Comment by mrocklin** [comment]
> Implement a parameter server for distributed training in dask-ml (see dask/dask-ml#171)

Whoa, that is definitely not a core issue.  I wouldn't put it on a Dask wish list.  It is important in the field of ML and deep learning though, so there is a sub-community that would care about it.  It is one of perhaps a hundred similar issues.

> If you just write down a few more items like this and put them on wiki or in the docs, this issue can be closed:)

I personally am unlikely to do this anytime in the next month (or longer?), though I encourage other people within the Dask community to collaborate and write down such a list that reflects common needs.

> I'm suggesting the Dask team does this, and writes down the outcomes. It's not hard, you already have talked to them a lot, and probably are aware of some things they want. Why let me reinvent the wheel?

If it was easy to do then I would be more than happy to do it.  The requests that come in are *very* broad.  I think that there is serious work to do here in collecting and analyzing any list that is shorter than 100 items long.  There are no wheels to reinvent here, so your help would be welcome.  

It's not that I'm withholding some information from you, it's that there is no good answer here.  If an institution like NumFocus wanted to help gather information here, then that would be welcome.  You may also be interested in https://github.com/dask/dask/issues/4748

**Comment by rgommers** [comment]
> Whoa, that is definitely not a core issue.

Okay. Since Machine Learning is pretty prominently in the doc navigation, that wasn't clear to me. The `dask` vs `dask-ml` distinction (and `distributed` etc) seems more like a project-internal thing with "Dask" being a project and a set of maintainers that work on the various repos. I'm interested in Dask the project more than dask the repo. Anyway, this can easily be addressed by a short statement on scope on any wishlist type doc.

> I personally am unlikely to do this anytime in the next month (or longer?), though I encourage other people within the Dask community to collaborate and write down such a list that reflects common needs.

Cool, happy to hear this. From the discussion so far it felt more like "we'd rather not have this".

> If it was easy to do then I would be more than happy to do it. The requests that come in are very broad. I think that there is serious work to do here in collecting and analyzing any list that is shorter than 100 items long. There are no wheels to reinvent here, so your help would be welcome.

My suggestion would be to be pragmatic. An incomplete list is better than no list. It's not necessary to hold a big user survey to start. Example: the NumPy and SciPy roadmaps were written by the core teams just by discussing on the mailing lists and in person. Those docs have been very helpful in different contexts. 

> It's not that I'm withholding some information from you, it's that there is no good answer here. If an institution like NumFocus wanted to help gather information here,

I'd be happy to help facilitate and do some prep work, and then perhaps have some higher bandwidth discussion at SciPy'19 or in a call before then?
